File: ts/11.binary-tree/is-binary-search-tree.ts
import { Node } from './';

export function depthFirstSearch<T extends number>(
    node: Node<T>,
    min: T,
    max: T
) {
    if (!node) {
        return true;
    }

    if (node.value > max || node.value < min) {
        return false;
    }

    if (node.right) {
        if (!depthFirstSearch(node.right, node.value, max)) {
            return false;
        }
    }

    if (node.left) {
        if (!depthFirstSearch(node.left, min, node.value)) {
            return false;
        }
    }

    return true;
}

export function isBinarySearchTree(root: Node<number> | null) {
    if (!root) return root;

    return depthFirstSearch<number>(root, -Infinity, +Infinity);
}

File: golang/8.binary-tree/main/is_binary_search_tree.go
package main

import (
	"github.com/AmirAhmadzadeh/problems/node"
	"math"
)

func isBinarySearchTreeCall(node *node.Node, min float64, max float64) bool {
	if node == nil {
		return true
	}

	if float64(node.Value()) < min || float64(node.Value()) > max {
		return false
	}

	if node.Right() != nil {
		if !isBinarySearchTreeCall(node.Right(), float64(node.Value()), max) {
			return false
		}
	}

	if node.Left() != nil {
		if !isBinarySearchTreeCall(node.Left(), min, float64(node.Value())) {
			return false
		}
	}

	return true
}

func isBinarySearchTree(root *node.Node) bool {
	return isBinarySearchTreeCall(root, math.Inf(-1), math.Inf(1))
}
